prediction/src/preprocess/lung_segmentation.py

```diff
@@ -0,0 +1,161 @@
+import logging
+import math
+import os
+import sys
+
+import cv2
+import dicom
+import numpy
+import scipy
+from skimage.filters import roberts
+from skimage.measure import label, regionprops
+from skimage.morphology import disk, binary_erosion, binary_closing
+from skimage.segmentation import clear_border
+from src.algorithms.identify.helpers import rescale_patient_images
+
+
+def get_z_range(dicom_path):
+    """Return the Z range of the images in the DICOM path
+    e.g. -379.0, -31.5
+
+    Args:
+        dicom_path: path that contains DICOM images
+
+    Returns:
+        minimum and maximum of the Z range
+    """
+    slices = load_patient(dicom_path)
+    min_z, max_z = sys.float_info.max, sys.float_info.min
+    for slice in slices:
+        z = float(slice.ImagePositionPatient[2])
+        if z < min_z:
+            min_z = z
+        if z > max_z:
+            max_z = z
+    return min_z, max_z
+
+
+def save_lung_segments(dicom_path, patient_id):
+    """Write the converted scan images and related lung masks to EXTRACTED_IMAGE_DIR.
+
+    Args:
+        dicom_path: a path to a DICOM directory
+        patient_id: SeriesInstanceUID of the patient
+
+    Returns:
+        shape of the original patient images (z, x, y),
+        shape of the rescaled mask images (z, x, y)
+    """
+    EXTRACTED_IMAGE_DIR = "data/extracted/"
+    TARGET_VOXEL_MM = 1.00
+    target_dir = os.path.join(os.getcwd(), EXTRACTED_IMAGE_DIR, patient_id)
+    os.makedirs(target_dir, exist_ok=True)
+
+    slices = load_patient(dicom_path)
+
+    cos_value = (slices[0].ImageOrientationPatient[0])
+    cos_degree = round(math.degrees(math.acos(cos_value)), 2)
+
+    original_image = get_pixels_hu(slices)
+    invert_order = slices[1].ImagePositionPatient[2] > slices[0].ImagePositionPatient[2]
+
+    pixel_spacing = slices[0].PixelSpacing
+    pixel_spacing.append(slices[0].SliceThickness)
+    image = rescale_patient_images(original_image, pixel_spacing, TARGET_VOXEL_MM)
+    if not invert_order:
+        image = numpy.flipud(image)
+
+    for index, org_img in enumerate(image):
+        patient_dir = target_dir
+        os.makedirs(patient_dir, exist_ok=True)
+        img_path = patient_dir + "img_" + str(index).rjust(4, '0') + "_i.png"
+        # if there exists slope,rotation image with corresponding degree
+        if cos_degree > 0.0:
+            org_img = cv_flip(org_img, org_img.shape[1], org_img.shape[0], cos_degree)
+        img, mask = get_segmented_lungs(org_img.copy())
+        org_img = normalize_hu(org_img)
+        cv2.imwrite(img_path, org_img * 255)
+        cv2.imwrite(img_path.replace("_i.png", "_m.png"), mask * 255)
+
+    return original_image.shape, image.shape
+
+
+def load_patient(src_dir):
+    slices = []
+    for s in os.listdir(src_dir):
+        try:
+            slices.append(dicom.read_file(os.path.join(src_dir, s)))
+        except Exception as e:
```

Comment:
This is the only that really caught my eye. Specific exceptions are usually best. But in this case, the error is being logged, so it isn't that big of a deal. If you twisted my arm, I'd probably say something like this may be a bit cleaner (if `SpecificException` is in fact the only expected exception).

```
try:
    slice = dicom.read_file(os.path.join(src_dir, s))
except SpecificException:
    logging.error("{} is not a valid DICOM".format(s))
else:
    slices.append(slice)
```

Reply:
Good point, thanks! I replaced the twisted code by your suggestion ;)